doc/monitoring_containers/monitoring_containers.md

```diff
@@ -0,0 +1,334 @@
+| Config DB    | SONiC Configuration Database |
+| CLI          | Command Line Interface       |
+
+# 1 Feature Overview
+SONiC is a collection of various switch applications which are held in docker containers
+such as BGP container and SNMP container. Each application usually includes several processes which are 
+working together to provide and receive the services from other modules. As such, the health of
+critical processes in each docker container is imperative not only for the docker
+container working correctly but also for the intended functionalities of entire SONiC switch.
+
+## 1.1 Monitoring
+This feature is used to monitor the running status of critical processes and critical resource 
+usage such as CPU, memory and disk of each docker container.
+
+We used Monit system tool to detect whether a critical process is running or not and whether 
+the resource usage of a docker container is beyond the pre-defined threshold.
+
+## 1.2 Auto-Mitigating
+This feature demonstrated docker container can be automatically shut down and
+restarted if one of critical processes running in docker container exits unexpectedly. Restarting
+the entire docker container ensures that configuration is reloaded and all processes in 
+docker container get restarted, thus increasing the likelihood of entering a healthy state.
+
+We leveraged the 'event listener' mechanism in supervisord to auto-restart a docker container
+if one of its critical processes exited unexpectedly. We also added a configuration option to make this 
+auto-restart feature dynamically configurable. Specifically users can run CLI to configure this 
+feature residing in Config_DB as enabled/disabled status.
+
+## 1.3 Requirements
+
+### 1.3.1 Functional Requirements
+1. Monit must provide the ability to generate an alert when a critical process has not
+    been alive for 5 minutes.
+2. Monit must provide the ability to generate an alert when the resource usage of
+    a docker container is larger than the pre-defined threshold.
+3. The event listener in supervisord must receive the signal when a critical process in 
+    a docker container crashed or exited unexpectedly and then restart this docker 
+    container.
+4. CONFIG_DB can be configured to enable/disable this auto-restart feature for each docker
+    container.. 
+5. Users can access the status of auto-restart feature via the CLI utility
+    1. Users can see current auto-restart status for docker containers.
+    2. Users can configure auto-restart status for a specific docker container.
+
+### 1.3.2 Configuration and Management Requirements
+Via the init_cfg.json file, these container features are disabled by default.
+Configuration of these features can be done via:
+1. config_db.json
+2. CLI
+
+### 1.3.3 Warm Reboot Requirements
+When switch reboots in the warm boot mode, auto-restart feature must ensure that systemd 
+service is stopped explicitly such that it will not affect warm reboot functionality. 
```

Comment:
This is not quite correct. It should read similar to,

```
During the fast-reboot/warm-reboot/warm-restart procedures in SONiC, a select number of critical
processes and the containers they reside in are stopped in a special manner (via signals or similar).
In this situations, we need to ensure these containers remain stopped until the reboot/restart procedure
is complete. Therefore, in order to prevent the auto-restart mechanism from restarting the containers
prematurely, it is the responsibility of the fast-reboot/warm-reboot/warm-restart procedure to explicitly
stop the systemd service which manages the container immediately after stopping and critical processes/container.
Once the systemd service is explicitly stopped, it will not attempt to automatically restart the container.
```

Reply:
Updated. Thanks for this clear explanation.